File: src/common/Codec.java
package common;

import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

/**
 * @see <a href="https://leetcode.com/problems/serialize-and-deserialize-binary-tree/#/description">Codec</a>
 */
public class Codec {
    private static final String spliter = ",";
    private static final String NN = "#";

    /**
     * Serialize.
     *
     * @param root treenode
     * @return string
     */
    public static String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb);
        return sb.toString();
    }

    // DFS
    private static void buildString(TreeNode treeNode, StringBuilder sb) {
        if (treeNode == null) {
            sb.append(NN).append(spliter);
        } else {
            sb.append(treeNode.val).append(spliter);
            buildString(treeNode.left, sb);
            buildString(treeNode.right, sb);
        }
    }

    /**
     * Deserialize.
     *
     * @param data string
     * @return treenode
     */
    public static TreeNode deserialize(String data) {
        Deque<String> nodes = new LinkedList<>();
        nodes.addAll(Arrays.asList(data.split(spliter)));
        return buildTree(nodes);
    }

    private static TreeNode buildTree(Deque<String> nodes) {
        String val = nodes.pop();
        if (val.equals(NN)) {
            return null;
        } else {
            TreeNode node = new TreeNode(Integer.valueOf(val));
            node.left = buildTree(nodes);
            node.right = buildTree(nodes);
            return node;
        }
    }
}

File: csharp/LeetCode/DataStructures/Codec.cs
namespace LeetCode.DataStructures
{
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    public class Codec
    {

        // Encodes a tree to a single string.
        public string serialize(TreeNode root)
        {
            var q = new Queue<TreeNode>();
            q.Enqueue(root);
            var list = new List<string>();
            while (q.Any())
            {
                var current = q.Dequeue();
                var val = GetString(current);
                list.Add(val);
                if (current != null)
                {
                    q.Enqueue(current.left);
                    q.Enqueue(current.right);
                }
            }
            var sb = new StringBuilder();
            sb.Append(list.First());
            for (var i = 1; i < list.Count; i++)
            {
                sb.Append($",{list[i]}");
            }
            return sb.ToString();
        }

        // Decodes your encoded data to tree.
        public TreeNode deserialize(string data)
        {
            var nodes = data.Split(',');
            var q = new Queue<TreeNode>();
            TreeNode root = null;
            if (nodes.Any())
            {
                root = GetNode(nodes.First());
                if (root != null)
                {
                    q.Enqueue(root);
                }
            }
            var i = 1;
            while (i < nodes.Length && q.Any())
            {
                var current = q.Dequeue();
                if (current != null)
                {
                    if (i < nodes.Length)
                    {
                        var left = GetNode(nodes[i++]);
                        current.left = left;
                        q.Enqueue(left);
                    }
                    if (i < nodes.Length)
                    {
                        var right = GetNode(nodes[i++]);
                        current.right = right;
                        q.Enqueue(right);
                    }
                }
            }
            return root;
        }

        private TreeNode GetNode(string s)
        {
            if (int.TryParse(s, out var val))
            {
                return new TreeNode(val);
            }
            else
            {
                return null;
            }
        }

        private string GetString(TreeNode n)
        {
            return n == null ? "#" : n.val.ToString();
        }
    }
}
